Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::StringBuilder

cdumez opened the pull request:
#### f4ca7c87e0ecbba8df2a92541089c0820f758227
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::StringBuilder
<a href="https://bugs.webkit.org/show_bug.cgi?id=283347">https://bugs.webkit.org/show_bug.cgi?id=283347</a>

Reviewed by Geoffrey Garen.

* Source/WTF/wtf/text/StringBuilder.cpp:
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::span8 const):
(WTF::StringBuilder::span16 const):
(WTF::StringBuilder::span const):
(WTF::StringBuilder::operator[] const):
(WTF::equal):

Canonical link: <a href="https://commits.webkit.org/286813@main">https://commits.webkit.org/286813@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ba7ea0987cf96b3d9aa18841c12d20b1e319b4ea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77140 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56175 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30055 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81732 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28422 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79257 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65323 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4471 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60460 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18507 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80207 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50404 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66216 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/40763 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47806 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23714 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26745 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/70325 "Found 16 jsc stress test failures: stress/attribute-custom-accessor.js.dfg-eager, stress/get-by-val-hoist-above-structure-2.js.eager-jettison-no-cjit, stress/get-by-val-hoist-above-structure-2.js.ftl-no-cjit-b3o0, stress/get-by-val-hoist-above-structure-2.js.ftl-no-cjit-no-put-stack-validate, stress/get-by-val-hoist-above-structure-2.js.ftl-no-cjit-small-pool ...") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68921 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24049 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83125 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/76418 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4520 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3045 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68733 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4676 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66189 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67978 "Found 1 new API test failure: /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/test-page-list (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11964 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10053 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/98671 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11948 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4466 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21588 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4486 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7921 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6245 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/text/StringBuilder.h:
```diff
@@ -87,9 +86,9 @@ class StringBuilder {
-    template<typename CharacterType> std::span<const CharacterType> span() const { return std::span(characters<CharacterType>(), length()); }
+    std::span<const LChar> span8() const { return unsafeMakeSpan(characters<LChar>(), length()); }
+    std::span<const UChar> span16() const { return unsafeMakeSpan(characters<UChar>(), length()); }
+    template<typename CharacterType> std::span<const CharacterType> span() const { return unsafeMakeSpan(characters<CharacterType>(), length()); }
```
Kind of sad that we have a span-compatible type internally (`m_string` / `m_buffer`), but `characters<T>()` converts it to raw pointer, and then we unsafely convert back to span here.

I wonder if we can invert the dependency. Move the logic from `characters<T>()` into `span<T>()`, and then have `characters<T>()` wrap `span<T>()` and return `.data()`. That way, we can get rid of the `unsafeMakeSpan()` calls.

cdumez replied:
Ok, I got rid of `characters<T>()` altogether.

geoffreygaren reviewed (approved):
r=me -- but I think this can be nicer

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::StringBuilder
https://bugs.webkit.org/show_bug.cgi?id=283347

Reviewed by Geoffrey Garen.

* Source/WTF/wtf/text/StringBuilder.cpp:
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::span8 const):
(WTF::StringBuilder::span16 const):
(WTF::StringBuilder::span const):
(WTF::StringBuilder::operator[] const):
(WTF::equal):

Canonical link: https://commits.webkit.org/286813@main
